extract.py

```diff
@@ -0,0 +1,161 @@
+from collections import defaultdict
+import difflib
+import json
+import logging
+from typing import Any, Dict, Generator, List, Tuple
+
+from dulwich.diff_tree import TreeChange
+from dulwich.repo import Repo
+from dulwich.walk import WalkEntry
+
+logger = logging.getLogger(__name__)
+
+
+def is_binary(repo: Repo, change: TreeChange) -> bool:
+    """
+    Checks if this blob is for binary file
+    :param repo: repository with change
+    :param change: single change between two trees
+    :return: does this blob contain non utf-8 characters
+    """
+    sha = (change.new.sha or change.old.sha)
+    try:
+        repo.get_object(sha).data.decode()
+    except UnicodeDecodeError as e:
+        return True
+    return False
+
+
+def into_lines(repo: Repo, sha: bytes) -> List[str]:
+    """
+    Get lines of blob
+    :param repo: repository of blob
+    :param sha: sha of blob
+    :return: list of lines in blob
+    """
+    return repo.get_object(sha).data.decode().splitlines()
+
+
+def calc_diff(repo: Repo, change: TreeChange) -> Tuple[int, int]:
+    """
+    Get number of added and deleted lines
+    :param repo: repository with change
+    :param change: single change between two trees
+    :return: number of lines added/deleted
+    """
+    diff_calc = difflib.Differ()
+    diffs = diff_calc.compare(into_lines(repo, change.old.sha),
+                              into_lines(repo, change.new.sha))
+    added = 0
+    deleted = 0
+    for diff in diffs:
+        if diff[0] == "+" and diff[1] != "+":
```

Comment:
please check that we don't lose any rows like `diff[0] == "+" and diff[1] == "+"`